File: app/src/test/java/com/github/goldy1992/mp3player/client/callbacks/subscription/MediaIdSubscriptionCallbackTest.java
package com.github.goldy1992.mp3player.client.callbacks.subscription;

import android.os.Handler;
import android.os.Looper;
import android.support.v4.media.MediaBrowserCompat.MediaItem;

import com.github.goldy1992.mp3player.client.MediaBrowserResponseListener;

import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.ArgumentCaptor;
import org.mockito.Captor;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;
import org.robolectric.RobolectricTestRunner;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertTrue;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.ArgumentMatchers.eq;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.never;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;
import static org.robolectric.Shadows.shadowOf;

@RunWith(RobolectricTestRunner.class)
public class MediaIdSubscriptionCallbackTest {

    private MediaIdSubscriptionCallback mediaIdSubscriptionCallback;

    @Mock
    private MediaBrowserResponseListener mediaBrowserResponseListener;

    private List<MediaItem> mediaItemList;

    private MediaItem mockMediaItem;

    @Captor
    ArgumentCaptor<ArrayList<MediaItem>> captor;

    @Before
    public void setup() {
        MockitoAnnotations.initMocks(this);
        Handler handler = new Handler(Looper.getMainLooper());
        this.mockMediaItem = mock(MediaItem.class);
        this.mediaIdSubscriptionCallback = new MediaIdSubscriptionCallback(handler);
        this.mediaItemList = Collections.singletonList(mockMediaItem);
    }

    @Test
    public void testRegisterListener() {
        final String key = "KEY";
       assertTrue(mediaIdSubscriptionCallback.getMediaBrowserResponseListeners().isEmpty());
       mediaIdSubscriptionCallback.registerMediaBrowserResponseListener(key, mediaBrowserResponseListener);
       // assert size is now 1
        final int restultSize = mediaIdSubscriptionCallback.getMediaBrowserResponseListeners().size();
        assertEquals(1, restultSize);
    }

    @Test
    public void testOnChildrenLoadedForSubscribedKey() {
        final String subscribedKey = "SubscribedKey";
        mediaIdSubscriptionCallback.registerMediaBrowserResponseListener(subscribedKey, mediaBrowserResponseListener);
        mediaIdSubscriptionCallback.onChildrenLoaded(subscribedKey, mediaItemList);
        shadowOf(Looper.getMainLooper()).idle();
        verify(mediaBrowserResponseListener, times(1)).onChildrenLoaded(eq(subscribedKey), captor.capture());
        ArrayList<MediaItem> children = captor.getValue();
        assertEquals(1, children.size());
        assertEquals(mockMediaItem, children.get(0));
    }

    @Test
    public void testOnChildrenLoadedForNonSubscribedKey() {
        final String subscribedKey = "SubscribedKey";
        final String nonSubscribedKey = "NonSubscribedKey";

        mediaIdSubscriptionCallback.registerMediaBrowserResponseListener(subscribedKey, mediaBrowserResponseListener);
        mediaIdSubscriptionCallback.onChildrenLoaded(nonSubscribedKey, mediaItemList);
        shadowOf(Looper.getMainLooper()).idle();
        verify(mediaBrowserResponseListener, never()).onChildrenLoaded(eq(nonSubscribedKey), any());
    }


}

File: app/src/test/java/com/github/goldy1992/mp3player/client/callbacks/subscription/MediaIdSubscriptionCallbackTest.kt
package com.github.goldy1992.mp3player.client.callbacks.subscription

import android.os.Handler
import android.os.Looper
import android.support.v4.media.MediaBrowserCompat
import com.github.goldy1992.mp3player.client.MediaBrowserResponseListener
import org.junit.Assert
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.*
import org.robolectric.RobolectricTestRunner
import org.robolectric.Shadows
import java.util.*

@RunWith(RobolectricTestRunner::class)
class MediaIdSubscriptionCallbackTest {
    private var mediaIdSubscriptionCallback: MediaIdSubscriptionCallback? = null
    @Mock
    private val mediaBrowserResponseListener: MediaBrowserResponseListener? = null
    private var mediaItemList: List<MediaBrowserCompat.MediaItem>? = null

    @Mock
    private var mockMediaItem: MediaBrowserCompat.MediaItem? = null
    @Captor
    var captor: ArgumentCaptor<ArrayList<MediaBrowserCompat.MediaItem>>? = null

    @Before
    fun setup() {
        MockitoAnnotations.initMocks(this)
        val handler = Handler(Looper.getMainLooper())
        mockMediaItem = Mockito.mock(MediaBrowserCompat.MediaItem::class.java)
        mediaIdSubscriptionCallback = MediaIdSubscriptionCallback(handler)
        mediaItemList = listOf(mockMediaItem!!)
    }

    @Test
    fun testRegisterListener() {
        val key = "KEY"
        Assert.assertTrue(mediaIdSubscriptionCallback!!.getMediaBrowserResponseListeners().isEmpty())
        mediaIdSubscriptionCallback!!.registerMediaBrowserResponseListener(key, mediaBrowserResponseListener)
        // assert size is now 1
        val restultSize = mediaIdSubscriptionCallback!!.getMediaBrowserResponseListeners().size
        Assert.assertEquals(1, restultSize.toLong())
    }

    @Test
    fun testOnChildrenLoadedForSubscribedKey() {
        val subscribedKey = "SubscribedKey"
        mediaIdSubscriptionCallback!!.registerMediaBrowserResponseListener(subscribedKey, mediaBrowserResponseListener)
        mediaIdSubscriptionCallback!!.onChildrenLoaded(subscribedKey, mediaItemList!!)
        Shadows.shadowOf(Looper.getMainLooper()).idle()
        Mockito.verify(mediaBrowserResponseListener, Mockito.times(1))!!.onChildrenLoaded(ArgumentMatchers.eq(subscribedKey), captor!!.capture())
        val children = captor!!.value
        Assert.assertEquals(1, children.size.toLong())
        Assert.assertEquals(mockMediaItem, children[0])
    }

    @Test
    fun testOnChildrenLoadedForNonSubscribedKey() {
        val subscribedKey = "SubscribedKey"
        val nonSubscribedKey = "NonSubscribedKey"
        mediaIdSubscriptionCallback!!.registerMediaBrowserResponseListener(subscribedKey, mediaBrowserResponseListener)
        mediaIdSubscriptionCallback!!.onChildrenLoaded(nonSubscribedKey, mediaItemList!!)
        Shadows.shadowOf(Looper.getMainLooper()).idle()
        Mockito.verify(mediaBrowserResponseListener, Mockito.never())!!.onChildrenLoaded(ArgumentMatchers.eq(nonSubscribedKey), ArgumentMatchers.any())
    }
}